kube/src/service/mod.rs

```diff
@@ -104,22 +103,40 @@ impl TryFrom<Config> for Service {
             .map_response(maybe_decompress)
             .into_inner();
 
-        let https: HttpsConnector<_> = config.try_into()?;
-        let mut connector = TimeoutConnector::new(https);
-        if let Some(timeout) = timeout {
-            // reqwest's timeout is applied from when the request stars connecting until
-            // the response body has finished.
-            // Setting both connect and read timeout should be close enough.
-            connector.set_connect_timeout(Some(timeout));
-            // Timeout for reading the response.
-            connector.set_read_timeout(Some(timeout));
-        }
-        let client: HyperClient<_, Body> = HyperClient::builder().build(connector);
+        let with_timeout = layer::layer_fn(|c| {
+            let mut connector = TimeoutConnector::new(c);
+            if let Some(timeout) = timeout {
+                // reqwest's timeout is applied from when the request stars connecting until
+                // the response body has finished.
+                // Setting both connect and read timeout should be close enough.
+                connector.set_connect_timeout(Some(timeout));
+                // Timeout for reading the response.
+                connector.set_read_timeout(Some(timeout));
+            }
+            connector
+        });
+
+        #[cfg(feature = "native-tls")]
+        let tls = connector::tls_connector(config.identity, config.root_cert, config.accept_invalid_certs)?;
+        #[cfg(feature = "rustls-tls")]
+        let tls = std::sync::Arc::new(connector::tls_config(
+            config.identity,
+            config.root_cert,
+            config.accept_invalid_certs,
+        )?);
+
+        #[cfg(not(any(feature = "proxy-native-tls", feature = "proxy-rustls-tls")))]
+        let conn = connector::https_connector(tls);
+        #[cfg(any(feature = "proxy-native-tls", feature = "proxy-rustls-tls"))]
+        let conn = connector::proxy_connector(tls, config.proxy_url.map(|s| s.parse()).transpose()?);
```

Comment:
Does this mean that we always use proxying if the compile time feature is on? That's a bit hairy for apps that just want to support it.

Reply:
I think we perhaps should do the feature checking within a match on `config.proxy_url` instead and emit a warning if `proxy_url` is set, but the feature missing.

Reply:
> Does this mean that we always use proxying if the compile time feature is on? 

No, `hyper-proxy` only proxies if the `url` [matches](https://github.com/tafia/hyper-proxy/blob/768fbf4c4596717548ae3cf6d4687f0a588b911f/src/lib.rs#L442). Otherwise, it just [calls the inner connector](https://github.com/tafia/hyper-proxy/blob/768fbf4c4596717548ae3cf6d4687f0a588b911f/src/lib.rs#L517-L518) (the same `https` connector). If `config.proxy_url` is `None`, no proxy is added to the `ProxyConnector`, so it does the latter.

I originally tried to make this optional layer (similar to `mayb_auth`), but couldn't make it compile. I can try again if that's worth it.

> I think we perhaps should do the feature checking within a match on `config.proxy_url` instead and emit a warning/error if `proxy_url` is set, but the feature missing.

Yeah, we should error when trying to connect with a config with `proxy_url` set, without the feature set.

Reply:
> If config.proxy_url is None, no proxy is added to the ProxyConnector

Ah ok. That's okay then. Layer wise, is that even  possible? I thought the Connector would be needed to build the Client, which needs to be passed to `.service`?

Reply:
`HttpConnector`, `HttpsConnector`, `TimeoutConnector`, and `ProxyConnector` are all `tower::Service<Uri>`, so we can compose them the same way as we're doing with  `tower::Service<Request<Body>>`. We're composing `TimeoutConnector` with `HttpsConnector` or `ProxyConnector`:

```rust
let conn = ServiceBuilder::new().layer(with_timeout).service(conn);
```

`HyperClient`'s builder's `build` method takes `Connect` which is a type alias of `tower::Service<Uri>`:

```rust
let client: HyperClient<_, Body> = HyperClient::builder().build(conn);
```

So, I thought it should be possible to make `ProxyConnector` into a layer like `TimeoutConnector` and make that an optional layer.

```rust
let conn = ServiceBuilder::new()
    .layer(with_timeout)
    .option_layer(maybe_proxy) // only if `config.proxy_url`
    .service(https_conn);
let client: HyperClient<_, Body> = HyperClient::builder().build(conn);
```

Reply:
Hmm, `<ProxyConnector as Service<Uri>>::Response` is not `Connection` (describes a type returned by a connector), so the above doesn't work. Might be worth opening a PR there.

The current implementation is good enough for now.

Reply:
Yeah, no need to overstretch. Service is new, it would be nice to have these pristine and exportable layers, but it's early days. I am happy to merge this more speculatively for now and wait to see what issues arise.